R/query_search.R

```diff
@@ -0,0 +1,467 @@
+##' the only supported namespaces are \code{tag} and \code{parameter}.
+##'
+##' @title Search for orderly reports matching criteria
+##'
+##' @param query The query string - see details and examples
+##'
+##' @param name Name of the report to search.  Only a single report
+##'   can be searched at once.
+##'
+##' @param draft Should draft reports be used searched?  This should
+##'   be used only in development.  Valid values are logical
+##'   (\code{TRUE}, \code{FALSE}) or use the string \code{newer} to
+##'   use draft reports where they are newer than archive reports.
+##'   For consistency, \code{always} and \code{never} are equivalent
+##'   to \code{TRUE} and \code{FALSE}, respectively.
+##'
+##' @param parameters Named list of parameters (as would be passed to
+##'   \code{\link{orderly_run}}) if your query uses parameters on the
+##'   right-hand-side of an expression.
+##'
+##' @inheritParams orderly_list
+##'
+##' @return A character vector of matching report ids, possibly
+##'   zero-length.  If the query is a "latest" query, then exactly one
+##'   report id, possibly NA.
+##'
+##' @export
+##' @examples
+##' # We need a few reports here to actually query.  There is a report in
+##' # the "demo" example called "other" that takes a parameter "nmin",
+##' # which is used to filter data - it's not terribly important what it
+##' # does here, but it can give us a set of reports to use.
+##'
+##' # The demo set also includes configuration for two tags, called
+##' # "dataset" and "plot" - the "dataset" tag will always be applied
+##' # as it is listed in the orderly.yml but we can still add the
+##' # "plot" tag interactively
+##' root <- orderly::orderly_example("demo")
+##'
+##' # A helper function to mass-produce reports will reduce noise a bit
+##' run1 <- function(nmin, tags = NULL) {
+##'   id <- orderly_run("other", root = root, echo = FALSE,
+##'                     parameters = list(nmin = nmin), tags = tags)
+##'   orderly_commit(id, root = root)
+##'   id
+##' }
+##'
+##' ids <- c(run1(0.1), run1(0.2, "plot"), run1(0.3))
+##'
+##' # We can then ask for all reports where the parameter nmin was more
+##' # than some value
+##' orderly::orderly_search("parameter:nmin > 0.15", "other", root = root)
+##'
+##' # Or use "&&" to find tags within a range
+##' orderly::orderly_search("parameter:nmin > 0.1 && parameter:nmin < 0.3",
+##'                         "other", root = root)
+##'
+##' # We can look for tags
+##' orderly::orderly_search("tag:plot", "other", root = root)
+##'
+##' # or exclude them
+##' orderly::orderly_search("!tag:plot", "other", root = root)
+##'
+##' # or combine that with the presence/absence of a tag
+##' orderly::orderly_search("parameter:nmin > 0.15 && !tag:plot",
+##'                         "other", root = root)
+##'
+##' # Use latest() over a query to find the latest report matching the
+##' # query expression.
+##' orderly::orderly_search("latest(parameter:nmin > 0.15)",
```

Comment:
You haven't got any examples of `[!]is.null(parameter:thing)` - but possibly not needed!